### DISCUSSION TITLE
Welcome to stable-diffusion-webui-wildcards-adetailer Discussions!

### DISCUSSION THREAD

**Discussion by uorufu** [uorufu/stable-diffusion-webui-wildcards-adetailer#1] [Announcements]
## 👋 Welcome!

Any questions or issues, let me know.

**Comment by Nziner** [comment]
Hi! Installed your updated version and it's like there's something not working in it.
1 screenshot - with your extension enabled
`Tried __0_ my_color__ eyes, and __my_color__ eyes,`
2 screenshot - with your extension turned off
`__my_color__ eyes,`
![cmd_SWvvArJZfE](https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer/assets/6568291/411267b6-727b-4c2f-87a7-ee581477860d)
![cmd_K8Plf6eZpd](https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer/assets/6568291/a70abc91-d744-4554-b31f-f0c3c0e87d31)

**Reply by uorufu** [reply]
Also make sure you have the latest. Not the first release. I still need to update that one. I tested the latest wildcard.py file extensively yesterday.

**Comment by Nziner** [comment]
Nah, it doesn't work. I have the latest version, added it via git pull.
![chrome_rYErSNgGnG](https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer/assets/6568291/b96d1f09-8d07-420c-9147-00d97fb07745)
![cmd_3K8i9W7BA5](https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer/assets/6568291/1947e12b-c143-477d-9319-ffef13d74039)
![notepad++_SANn8BjED3](https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer/assets/6568291/cb492225-a2b3-4042-9214-31b72b184826)
![notepad++_mZgv750Vbi](https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer/assets/6568291/2285d211-9da2-4316-8b54-60a34ecc6ec2)

**Reply by uorufu** [reply]
You are trying to nest the wildcards, those are very difficult to get working in conjunction with adetailer.

you can add something like this:

 ```
                                 text[i] = self.replace_wildcard(line, rand_list_tiers[x], replacement_file)
                     nested = 0
                     if "__" in text[i]:
                         nestedtext = text[i].split("__")
                         nestedresult = " "
                         while k < len(nestedtext):
                                 nestedline = str(nestedtext[k])
                                 replacement_file = os.path.join(wildcards_dir, f"{nestedline}.txt")
                                 if os.path.exists(replacement_file):
                                     if nested > 99:
                                         nested = nested % 99
                                     nestedtext = self.replace_wildcard(nestedline, rand_list[nested], replacement_file)
                                     nestedresult = nestedresult + nestedtext
                                     nested = nested + 1
                             k = k + 1
                         text[i] = nestedresult
                     if len(p.all_seeds) > 1:
 ```

 inbetween line 142 and 143

Basically this checks for the existance of __ when normally they shouldnt exist anymore unless __ was pulled from txt files. Then it will give them random seeds according to how many nested wildcards are pulled from the text file.

So for example if you have a txt file like this

line 1: __wildcard1__ __wildcard2__ 
line 2: __wildcard3__ 

Then 1 and 3 will always have the same random seed (so if wildcard1 and 3 have equal lines, it will pull the same line from both. While wildcard2 is on a different seed.)

This is probably the best way to make it work with adetailer, so you dont get different results in adetailer. 

Personally, using nested wildcards sounds like a nightmare to handle to me, so I am not adding it to the main script and I havent tested this code, although it should work. But I personally would advice not using nested wildcards. Because adetailer prompts are generated and processed separately it is very finicky when it comes to matching the wildcards while keeping decent randomness intact. 

Also, this will only go one deep on nested results. If you nest wildcard continiously, you would have to paste this code inside multiple times.

 ```
                                 text[i] = self.replace_wildcard(line, rand_list_tiers[x], replacement_file)
                     nested = 0
                     if "__" in text[i]:
                         nestedtext = text[i].split("__")
                         nestedresult = " "
                         while k < len(nestedtext):
                                 nestedline = str(nestedtext[k])
                                 replacement_file = os.path.join(wildcards_dir, f"{nestedline}.txt")
                                 if os.path.exists(replacement_file):
                                     if nested > 99:
                                         nested = nested % 99
                                     nestedtext = self.replace_wildcard(nestedline, rand_list[nested], replacement_file)
                                     nestedresult = nestedresult + nestedtext
                                     nested = nested + 1
                             k = k + 1
                         text[i] = nestedresult
                     nested = 0
                     if "__" in text[i]:
                         nestedtext = text[i].split("__")
                         nestedresult = " "
                         while k < len(nestedtext):
                                 nestedline = str(nestedtext[k])
                                 replacement_file = os.path.join(wildcards_dir, f"{nestedline}.txt")
                                 if os.path.exists(replacement_file):
                                     if nested > 99:
                                         nested = nested % 99
                                     nestedtext = self.replace_wildcard(nestedline, rand_list[nested], replacement_file)
                                     nestedresult = nestedresult + nestedtext
                                     nested = nested + 1
                             k = k + 1
                         text[i] = nestedresult
                     if len(p.all_seeds) > 1:
 ```

For example means you can nest wildcards inside of another wildcard. 
If you dont care about matching wildcards with adetailer, there are simpler ways, but if you want to keep the randomness somewhat predictable and matchable between main and adetailer prompts, there isn't really a good way to do so I think.

**Reply by Nziner** [reply]
I think the problem here is something else. I don't even test with adetailer, I just write a normal prompt `__0_test1__`
And it gives the result `__0_test1__`

Now I tested the [first version](https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer/blob/b220cdad2fa9bc31d3d37696ca882dff2f29ba9e/scripts/wildcards.py), it at least processes the primary file and gives the result `__test3__`

File test1.txt:
```
__test2__
__test3__
```

File test2.txt:
green
purple

File test3.txt:
red
blue

**Comment by uorufu** [comment]
> I think the problem here is something else. I don't even test with adetailer, I just write a normal prompt `__0_test1__` And it gives the result `__0_test1__`
> 
> Now I tested the [first version](https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer/blob/b220cdad2fa9bc31d3d37696ca882dff2f29ba9e/scripts/wildcards.py), it at least processes the primary file and gives the result `__test3__`
> 
> File test1.txt:
> 
> ```
> __test2__
> __test3__
> ```
> 
> File test2.txt: green purple
> 
> File test3.txt: red blue

Honestly I don't understand why it isn't working for you then. I just made the same txt files you made and tested it for normal, iterative and tiered wildcards and it works fine.
The script as is the results become either `__test2__` or `__test3__` 
and with this added part (I forgot one thing and it had bad indentation before apparantly, but it works with this:
```
                    nested = 0
                    if "__" in text[i]:
                        nestedtext = text[i].split("__")
                        nestedresult = " "
                        k=0
                        while k < len(nestedtext):
                            nestedline = str(nestedtext[k])
                            replacement_file = os.path.join(wildcards_dir, f"{nestedline}.txt")
                            if os.path.exists(replacement_file):
                                if nested > 99:
                                    nested = nested % 99
                                nestedtext = self.replace_wildcard(nestedline, rand_list[nested], replacement_file)
                                nestedresult = nestedresult + nestedtext
                                nested = nested + 1
                            k = k + 1
                        text[i] = nestedresult
                    nested = 0
                    if "__" in text[i]:
                        nestedtext = text[i].split("__")
                        nestedresult = " "
                        k=0
                        while k < len(nestedtext):
                            nestedline = str(nestedtext[k])
                            replacement_file = os.path.join(wildcards_dir, f"{nestedline}.txt")
                            if os.path.exists(replacement_file):
                                if nested > 99:
                                    nested = nested % 99
                                nestedtext = self.replace_wildcard(nestedline, rand_list[nested], replacement_file)
                                nestedresult = nestedresult + nestedtext
                                nested = nested + 1
                            k = k + 1
```

I get the colors properly from text2.txt and text3.txt

Here you see it first gets the lines from test1 and then the line from test2 or 3 

![Screenshot 2024-04-16 120844](https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer/assets/153657365/92d5981e-8b02-47d2-aea1-b52b1f37cc5c)

**Reply by Nziner** [reply]
The second section of code I forgot to try. It really works!

I just need to make it possible to use multiple wildcards or words after.
![cmd_dOrHkzeORv](https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer/assets/6568291/482f9904-f7de-46be-b53b-b05837028ea0)

**Comment by Bra2ha** [comment]
Hi, does your current version of wildcards.py correctly apply per-image hires prompts when batch count > 1 with Hires Fix enabled?

**Reply by uorufu** [reply]
> Hi, does your current version of wildcards.py correctly apply per-image hires prompts when batch count > 1 with Hires Fix enabled?

Yes, using latest forge you should have no issues. 
Only thing is it doesn't save properly in image meta when it comes to hires, if you save image before hires as well, you'll see the prompts (including hires prompt) get saved properly in the image before hires for each image in the batch, but not in the hires result. hires image always saves the prompts from the first image results for some reason in forge. But the generation process does apply the correct prompt for each image.

**Reply by Bra2ha** [reply]
> hires image always saves the prompts from the first image results for some reason in forge.

This is fixed in my version of wildcards.py. Take a look if you interested. 

It also has 4 modes:
2 underscores - original random.
3 underscores - sequential.
4 underscores -  random, no repeats in a batch.
5 underscores -  random, no repeats in a session.

Also you can use "_" in wildcard file names (for example, colors_light.txt).
[wildcards.py](https://github.com/user-attachments/files/23698967/wildcards.py)
